packages/zoltan2/src/input/Zoltan2_Adapter.hpp

```diff
@@ -116,22 +117,31 @@ template <typename User>
   virtual ~BaseAdapter() {};
 
   /*! \brief Provide a pointer to this process' identifiers.
-
       \param ids will on return point to the list of the global Ids for 
         this process.
    */
   virtual void getIDsView(const gno_t *&ids) const {
-    Kokkos::View<gno_t *> kokkosIds;
+    Kokkos::View<const gno_t *, Kokkos::HostSpace> kokkosIds;
     getIDsKokkosView(kokkosIds);
     ids = kokkosIds.data();
   }
 
   /*! \brief Provide a pointer to this process' identifiers.
-
       \param ids will on return point to the list of the global Ids for 
         this process.
    */
-  virtual void getIDsKokkosView(Kokkos::View<gno_t *> &/* ids */) const {
+  virtual void getIDsKokkosView(Kokkos::View<const gno_t *,
+    Kokkos::HostSpace> &ids) const {
+    Z2_THROW_NOT_IMPLEMENTED
+  }
+
+  /*! \brief Provide a pointer to the coordinates view.
+  */
+  inline void getCoordinatesKokkosView(
+    // coordinates in MJ are LayoutLeft since Tpetra Multivector gives LayoutLeft
```

Comment:
@kddevin I think a different interface could help.  For example (requires C++14 but easy to work around in legal C++11):
```c++
withCoordinates (
  [] (const auto& coords) { /* user's code that accesses coords */ },
  readOnly (zoltan2AdapterInstance).as (Kokkos::LayoutLeft ()).on (Kokkos::HostSpace ())
);
```
This says, "Within the given lambda or function, I want to access the coordinates (local to my MPI process) in read-only fashion, as a LayoutLeft Kokkos::View in host memory."  If the coordinates are stored as a LayoutLeft View in host memory, then I can just view them.  Otherwise, the implementation of `withLocalCoordinates` would make a temporary copy.  `readOnly` tells Zoltan2 that it need not copy the coordinates back to their original storage format after use. 

You could have different `with*` functions for different kinds of data: coordinates, weights, global indices, etc.  Alternately, you could specify the kind of data by tag:
```c++
withLocalDataAccess (
  [] (auto coords) { /* user's code that accesses coordinates */ },
  from (zoltan2AdapterInstance).readOnly (coordinates_tag).as (Kokkos::LayoutLeft ()).on (Kokkos::HostSpace ())
);

withLocalDataAccess (
  [] (auto weights) { /* user's code that accesses weights */ },
  from (zoltan2AdapterInstance).readWrite (weights_tag).as (Kokkos::LayoutRight ()).on (Kokkos::CudaSpace ())
);
```
With this tag-based approach, you could access multiple kinds of data at a time:
```c++
withLocalDataAccess (
  [] (auto coords, auto weights) { /* user's code */ },
  from (zoltan2AdapterInstance).readOnly (coordinates_tag).as (Kokkos::LayoutLeft ()).on (Kokkos::HostSpace (),
  from (zoltan2AdapterInstance).readWrite (weights_tag).as (Kokkos::LayoutRight ()).on (Kokkos::CudaSpace ()
);
```

Reply:
Currently we have MJ fixed to use Layout Left for the coordinates since we had decided we wanted to be able to use the MultiVector coordinates directly without a copy, which is currently the case. The only reason for LayoutLeft appearing in a few places in MJ is because of this. If we did decide to copy they could all go away, though we do take subviews of the coordinates since MJ likes to work with coords just on a single dimension and we have dimension as the 2nd parameter. I'm not sure how it would affect performance to change these things. Other adapters create a LayoutLeft coordinate view as well, but this is only for compatibility with MultiVector which currently determines MJ's layout type.

So we have this flow:
  MultiVector is LayoutLeft
  So we made MJ coords LayoutLeft
  So we made other adapter coords LayoutLeft